compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -602,7 +607,53 @@ public void visit(BLangObjectTypeNode objectTypeNode) {
     @Override
     public void visit(BLangRecordTypeNode recordTypeNode) {
         SymbolEnv recordEnv = SymbolEnv.createTypeEnv(recordTypeNode, recordTypeNode.symbol.scope, env);
-        recordTypeNode.fields.forEach(field -> analyzeDef(field, recordEnv));
+
+        BType type = recordTypeNode.type;
+
+        boolean isRecordType = false;
+        LinkedHashMap<String, BField> fields = null;
+
+        boolean allReadOnlyFields = false;
+
+        if (type.tag == TypeTags.RECORD) {
+            isRecordType = true;
+
+            BRecordType recordType = (BRecordType) type;
+            fields = recordType.fields;
+            allReadOnlyFields = recordType.sealed;
+        }
+
+        for (BLangSimpleVariable field : recordTypeNode.fields) {
+            if (field.flagSet.contains(Flag.READONLY)) {
+                BType fieldType = field.type;
+
+                if (fieldType == symTable.semanticError) {
+                    continue;
+                }
+
+                BType readOnlyFieldType = getReadOnlyFieldType(field.pos, fieldType);
+
+                if (readOnlyFieldType == symTable.semanticError) {
+                    dlog.error(field.pos, DiagnosticErrorCode.INVALID_READONLY_FIELD_TYPE, fieldType);
+                } else {
+                    if (isRecordType) {
+                        fields.get(field.name.value).type = readOnlyFieldType;
+                    }
+
+                    field.type = field.symbol.type = readOnlyFieldType;
+                }
+            } else {
+                allReadOnlyFields = false;
+            }
```

Comment:
Can we extract a method here

Reply:
Moved the logic in the if block to a separate method in dd6fc00.